### ISSUE TITLE
Broken decontamination shower

### ISSUE THREAD

**Issue by TenAltz** [CorsixTH/CorsixTH#1431]
Hi, Im on a high level, and suddently the game freezes. 
The patients and staff are not moving, changing the gamespeed does nothing. When I try another save game before the crash, it still wont move at all. 

If I restart the level, my screen goes black until I click and the boardgame comes up. Then I manually have to click my way through the cutscene one frame pr. click. When the game finally restarts, I can build fine, but the game is not reacting to different game-speeds. (no sound or announcements)

I have Copy/pasted the eventlog:

An error has occurred!
Almost anything can be the cause, but the detailed information below can help the developers find the source of the error.
Running: The timer handler.
A stack trace is included below, and the handler has been disconnected.
Lua\entities\humanoid.lua:868: Action queue was empty. This should never happen.
humanoid - class: Standard Male Patient
Warmth: 0.210   Happiness: 0.997   Fatigue: 0.000  Thirst: 0.113  Toilet_Need: 0.366   Health: 0.000
Actions: []
stack traceback:
        [C]: in function 'error'
        Lua\entities\humanoid.lua:868: in method 'getCurrentAction'
        Lua\epidemic.lua:550: in upvalue 'is_static'
        Lua\epidemic.lua:561: in method 'markedPatientsCallForVaccination'
        Lua\epidemic.lua:109: in method 'tick'
        Lua\hospital.lua:1385: in method 'manageEpidemics'
        Lua\hospital.lua:822: in method 'tick'
        Lua\world.lua:1023: in method 'onTick'
        Lua\app.lua:1014: in function <Lua\app.lua:1011>
        (...tail calls...)
        Lua\app.lua:915: in function <Lua\app.lua:910>

Warning: No event handler for timer #### 

#### Steps to Reproduce
1. *Load save game attached to this ticket*
https://www.dropbox.com/s/f481qxm4o968az9/TENNA.H8?dl=0
https://www.dropbox.com/s/o03uol0aejl3kz5/TENNA.H99?dl=0
https://www.dropbox.com/s/4q771xob8ori3ws/TENNA.H1?dl=0

^ Link to my recent savegames

#### System Information
**CorsixTH Version:  0.62

**Operating System: Windows 10*

**Theme Hospital Version: Origin

**Comment by TheCycoONE** [comment]
Does this happen when you start CorsixTH again?

**Comment by TenAltz** [comment]
When I start CortixTH (i have done this at least 10 times over a couple of days) and press resumé game, the game is still frozen along with my other saves (i have savegames from early levels, and they dont move any more)

If I try to start completely over, the game still doesent move, when I change the gamespeed. 

Furthermore when I go back to the main menu when I have tried to play one of the levels, my mouse only moves when clicked.

**Comment by TheCycoONE** [comment]
Those are Theme Hospital save files, not CorsixTH saves.

**Comment by mugmuggy** [comment]
That helps. Patient using the decontamination shower is 'going home' for some reason, but the room can't finalise the processing and make him leave in that instance.

https://github.com/CorsixTH/CorsixTH/blob/b8e9d7ff500f357b8e7a6562e4aaec22c48ee102/CorsixTH/Lua/rooms/decontamination.lua#L106-L108

**Comment by TheCycoONE** [comment]
So on a brand new game (not loading the save) speed works correctly or are there two issues?

**Comment by mugmuggy** [comment]
Brand new process, new game would work fine.
New process, load bad save game, exit that game/restart/load another working save (as its not working) - new game not working.

Main menu can get a choppy - hence the: Then I manually have to click my way through the cutscene one frame pr. click.

Its hard to guess the state before the save, but as the patient.going_home flag is true, the startAction handler for an empty action queue will not run either.

https://github.com/CorsixTH/CorsixTH/blob/b8e9d7ff500f357b8e7a6562e4aaec22c48ee102/CorsixTH/Lua/entities/humanoid.lua#L462-L467

But in this case its like it already has done the startAction previously, kicked the patient but ended up wiht an empty queue again, though the patient does only appear to be finishing using the shower on loading the 6 savegame.

General diagnosis for example handles soon to be empty action queues:
https://github.com/CorsixTH/CorsixTH/blob/b8e9d7ff500f357b8e7a6562e4aaec22c48ee102/CorsixTH/Lua/rooms/general_diag.lua#L76-L80

I hacked in an idle action and then followed up with a dealtWithPatient call through debug scripts and the game continues as normal.

But knowing why the whole system appears to breakdown at this point, would be good to know also, I haven't dipped my feet in on that side that much yet. But suspect that the error processing isn't getting an entity passed to it, to turn off the 'ticks' of the patient leaving the shower. Re-enabling the timer, and then disabling ticks on the patient and things 'resume' there also, just with the patient broken. Ie know recovery window pops up as:
https://github.com/CorsixTH/CorsixTH/blob/b8e9d7ff500f357b8e7a6562e4aaec22c48ee102/CorsixTH/Lua/app.lua#L961

self.world.current_tick_entity is null and the handler isn't ever reattached, leaving eventHandlers["timer"] = nil and therefore no app:onTick() calls are made. Or something like that.

**Comment by TenAltz** [comment]
A new game works fine. I just dont want to start all over again unless that is the only option. 

> That helps. Patient using the decontamination shower is 'going home' for some reason, but the room can't finalise the processing and make him leave in that instance.

I think I know what the problem to that is. I just now saw that, when I tried to load the broken save: 
https://imgur.com/xuYMHMt 

So... what do I do now? Im not some kind of computer/programmer expert - I just want to play my game. :-)

**Comment by mugmuggy** [comment]
After some thought and a bit of a trial and error I was able to manipulate a game state into replicating this issue. The 'easiest' way to replicate is to have a patient waiting for the decontamination room to be built (ie they can't be in a queue).

That patients health has to be very low. The patient also has to be thirsty enough so the drinks machine will be accessed (you also may need math.random luck, on your side). The patient:tickDay is called, reducing the health to 0, and also sending the patient to the drinks machine.

The walk action to drinks machine is
```lua
self:queueAction(WalkAction(lx, ly):setMustHappen(true):disableTruncate(), 1)
```
Whilst the patient is then routed to the drinks machine, the patient tickDay is called again (so patient speed, plus drinks machine distance helps here). That then leads to tickDay calling Patient:die, setting the patient.go_home to true and queuing the meander and die actions. The setNextAction for the patient doesn't interrupt the walk to drinks machine due to the disabled truncate.

Player must also simultaneously build the room (this could be completing an edit whilst all this is going on) either a callback is registered already and exists or the seek_room action completes, leading to a walk to the room, interrupting the meander and die actions.

Thus allowing a dead patient to attempt to enter a room again and as patient.go_home is already set to true, they will not tick again, removing any other opportunity for them to die again and will enter the room and complete actions as they can.

Patient:die probably should call Patient:unregisterCallbacks as they don't care at that point, rather than waiting for the despawn to prevent seek_room callbacks that exist, if any from running, but also that walk with disableTruncate appears to be a factor and likewise dealingWithPatient when queue is about to be empty would also bypass this issue.

**Comment by TheCycoONE** [comment]
I like the idea of unregistering callbacks on die(). I didn't entirely follow `dealingWithPatient`

**Comment by mugmuggy** [comment]
DealingWithPatient from my earlier comment, don't test patient.going_home but length of the action_queue instead, so dealingWithPatient is called to insert new actions.
https://github.com/CorsixTH/CorsixTH/blob/b8e9d7ff500f357b8e7a6562e4aaec22c48ee102/CorsixTH/Lua/rooms/decontamination.lua#L106-L108

The scanner room also has this same fault checking going_home instead of the action_queue length, when this incorrect state occurs.

It also doesn't have to be as complicated to get it to occur, just most rooms are protecting against it I guess, seek room action starts, creates callbacks, that inserts a meander action at start of queue, which then inserts a walk at start of queue, the walk starts, patient tickDay calls patient die, and room is built., but again timing has to be pretty tight (build while paused or opening the room between game world ticks).

Also probably have to TheApp.ui.bottom_panel:removeMessage(self) in patient:die if removing the callbacks otherwise that will linger. I'll test it out when I have some spare time.

**Comment by lewri** [comment]
May be resolved by PR #1579

**Comment by TheCycoONE** [comment]
@mugmuggy can you verify ^

**Comment by mugmuggy** [comment]
@TheCycoOne, indeed when this popped up in a test on another hospital, #1579 resolves the stack trace instance, alas like #1584 it doesn't cover all instances of a dying patient as there is a timing issue. Happy to resolve this in favour of #1584 for the other possibility.
